Mathlib/Data/Pi/Algebra.lean

```diff
@@ -0,0 +1,398 @@
+theorem apply_mul_single (f' : ∀ i, f i → g i) (hf' : ∀ i, f' i 1 = 1) (i : I) (x : f i) (j : I) :
+    f' j (mulSingle i x j) = mulSingle i (f' i x) j := by
+  simpa only [Pi.one_apply, hf', mulSingle] using Function.apply_update f' 1 i x j
+
+@[to_additive apply_single₂]
+theorem apply_mul_single₂ (f' : ∀ i, f i → g i → h i) (hf' : ∀ i, f' i 1 1 = 1) (i : I)
+    (x : f i) (y : g i) (j : I) :
+    f' j (mulSingle i x j) (mulSingle i y j) = mulSingle i (f' i x y) j := by
+  by_cases h : j = i
+  · subst h
+    simp only [mul_single_eq_same]
+
+  · simp only [mul_single_eq_of_ne h, hf']
+
+@[to_additive]
+theorem mul_single_op {g : I → Type _} [∀ i, One (g i)] (op : ∀ i, f i → g i)
+    (h : ∀ i, op i 1 = 1) (i : I) (x : f i) :
+    mulSingle i (op i x) = fun j => op j (mulSingle i x j) :=
+  Eq.symm <| funext <| apply_mul_single op h i x
+
+@[to_additive]
+theorem mul_single_op₂ {g₁ g₂ : I → Type _} [∀ i, One (g₁ i)] [∀ i, One (g₂ i)]
+    (op : ∀ i, g₁ i → g₂ i → f i) (h : ∀ i, op i 1 1 = 1) (i : I) (x₁ : g₁ i) (x₂ : g₂ i) :
+    mulSingle i (op i x₁ x₂) = fun j => op j (mulSingle i x₁ j) (mulSingle i x₂ j) :=
+  Eq.symm <| funext <| apply_mul_single₂ op h i x₁ x₂
+
+variable (f)
+
+@[to_additive]
+theorem mul_single_injective (i : I) : Function.Injective (mulSingle i : f i → ∀ i, f i) :=
+  Function.update_injective _ i
+
+@[simp, to_additive]
+theorem mul_single_inj (i : I) {x y : f i} : mulSingle i x = mulSingle i y ↔ x = y :=
+  (Pi.mul_single_injective _ _).eq_iff
+
+end
+
+/-- The mapping into a product type built from maps into each component. -/
+@[simp]
+protected def prod (f' : ∀ i, f i) (g' : ∀ i, g i) (i : I) : f i × g i :=
+  (f' i, g' i)
+
+-- Porting note : These two are simp-normal-form versions of the lemmas below.
+-- They should probably be placed in a lower file and be renamed.
+@[simp]
+theorem prod_fst_snd₂ : (fun (i : α × β) => i) = id :=
+  funext fun _ => Prod.mk.eta
+
+@[simp]
+theorem prod_snd_fst₂ : (fun (i : α × β) => (i.snd, i.fst)) = Prod.swap :=
+  rfl
+
+-- Porting note : Linter says these two are not in simp-normal-form. Above is the normal
+-- form version.
+-- @[simp]
+theorem prod_fst_snd : Pi.prod (Prod.fst : α × β → α) (Prod.snd : α × β → β) = id :=
+  funext fun _ => Prod.mk.eta
+
+-- @[simp]
+theorem prod_snd_fst : Pi.prod (Prod.snd : α × β → β) (Prod.fst : α × β → α) = Prod.swap :=
+  rfl
+
+end Pi
+
+namespace Function
+
+section Extend
+
+@[to_additive]
+theorem extend_one [One γ] (f : α → β) : Function.extend f (1 : α → γ) (1 : β → γ) = 1 :=
+  funext fun _ => by apply ite_self
+
+-- Porting note : Had to add `[∀ x, Decidable (∃ a, f a = x)]` to these
+-- next 3 theorems to proof them, but not sure that's correct.
+@[to_additive]
+theorem extend_mul [Mul γ] (f : α → β) (g₁ g₂ : α → γ) (e₁ e₂ : β → γ)
+    [∀ x, Decidable (∃ a, f a = x)]:
+    Function.extend f (g₁ * g₂) (e₁ * e₂) = Function.extend f g₁ e₁ * Function.extend f g₂ e₂ := by
+  funext x
+  simp [Function.extend_def, apply_dite₂]
```

Comment:
I did not manage a term proof and used tactic mode instead. Is that okay?

Reply:
Also I do not know if adding `[∀ x, Decidable (∃ a, f a = x)]` will have any concequences or not.

Reply:
First: you should instead `import Mathlib.Tactic.Classical` and use `classical` as the first step of the proof, rather than adding decidability instances as hypotheses. (We only use decidable in hypotheses when they are used in the statement.)

Second: I'd like to understand why the original proof using `convert` fails.

Reply:
Removed the `[.. Decidable ..]` as suggested.

As for the Second point, I'm afraid I don't think I have the understanding to figure that out, but if anybody else wants to investigate, the previous proofs are in the comment.